src/core/thread/mesh_forwarder.cpp

```diff
@@ -2016,6 +2016,11 @@ void MeshForwarder::HandleDataRequest(const Mac::Address &aMacSource, const Thre
     child->SetLastHeard(Timer::GetNow());
     child->ResetLinkFailures();
 
+    if (aMacSource.mLength == sizeof(aMacSource.mShortAddress))
+    {
+        mSourceMatchController.SetSrcMatchAsShort(*child, true);
```

Comment:
One downside with this approach is that this received Data Request message is wasted and cannot trigger delivery of a pending Data Frame.

A more effective approach might be to simply enable short source address matching after the first Indirect Transmission attempt to the child.  This will result in the following two cases:
1. The Child ID Response made it to the child.  The child will use its short address in the next Data Request frame.
2. The Child ID Response did not make it to the child.  The child will need to send a new Child ID Request anyway.

In this way, we never have a situation where we "waste" a Data Request message.